### ISSUE TITLE
QUESTION: Help with typescript for atom package

### ISSUE THREAD

**Issue by DavidLGoldberg** [TypeStrong/atom-typescript#1362]
Hi, (feel free to "close" this issue ASAP, just a question)

Developer of Jumpy here (thanks for listing my package in your favorite packages!)
I'm actually looking to upgrade Jumpy to -> JS -> TS....and well actually create a core engine that's in Elm, so Elm would interface with TS instead of the more typical Elm interfacing with JS but that's another story!

As a baby step:
I have a very small file in a new branch of mine that's 5 lines, converted from CoffeeScript into JS and then now into TS.  It compiles (and works) and your package is already giving me awesome TS info etc.  I am currently using your package to auto save my typescript and write to a /dist just like you guys are.

I had to copy your package.json I ended up with (look good to you guys?):

```
  "dependencies": {
...
    "tsconfig": "6.0.0",
    "tslib": "1.7.1",
    "typescript": "2.5.3"
  },
  "devDependencies": {
    "@types/atom": "0.0.38",
    "ts-node": "3.2.1"
  }
```

and for the next person, I also put a tsconfig.json in my /lib folder
```
{
  "compilerOptions": {
    "forceConsistentCasingInFileNames": true,
    "importHelpers": true,
    "lib": ["es6", "dom"],
    "module": "commonjs",
    "noLib": false,
    "outDir": "../dist",
    "preserveConstEnums": true,
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true,
    "target": "es6"
  },
  "compileOnSave": true
}
```

I was wondering though...the JS works with just a rename to .js and 'use babel';
Kind of a shame we have to commit our /dist folders right?

This is kind of the same issue I saw with the way someone was doing an elm package as well.

Basically, if apm publish always ran some npm build of sorts we wouldn't have to rely on committing /dist right?

Just curious about everyone's thoughts and the current status of all of this.

P.S. Prior to copying your /dist I tried just renaming and it looked like Atom *was* trying to compile it to typescript but ran into issues.  The github issues I could find there were kind of a rabbit hole (https://github.com/smhxx/atom-ts-transpiler/issues/8#issuecomment-334346709).  I wonder if we could just do one better by way of an npm step?

So atom-typescript for auto build local, npm step for when published..... although they'd better be taking the same paths ;)

**Comment by lierdakil** [comment]
> I had to copy your package.json I ended up with (look good to you guys?):

You don't need either `tsconfig` or `tslib` unless you're interfacing with those directly in your code. You can also move `typescript` to `devDependencies`, since you're transpiling to JS before publishing (at least I hope you are) -- no need to force users to install typescript.

Also, you don't really need `ts-node` unless you're writing specs in TS or using it to directly run some TS scripts on CI (or somewhere else).

By the way, and if you are writing specs in TS, you'll need something akin to https://github.com/TypeStrong/atom-typescript/pull/1353

I should also mention that `@types/atom` is horribly outdated. I've slapped together a more up-to-date version at one point, which you can find, among other things, [here](https://github.com/atom-haskell/typings), but some definitions aren't as good as they could be (those are mostly generated based on Atom's docs and then painstakingly fixed by hand, because Atom's docs aren't the best they could be either).

> Basically, if apm publish always ran some npm build of sorts we wouldn't have to rely on committing /dist right?

Actually, no. Atom packages are little more than git tags. So if you don't have something in git, it means it needs to happen on user-side. For instance, Atom automatically transpiles coffeescript and ESnext (i.e. JS with `use babel` annotation) on first use. 

Atom does have some support for typescript actually, but it uses ancient typescript version, and isn't suitable for general use at all. And since apparently Atom bet on Facebook's flow (which is supported out-of-the-box with an annotation), I doubt there's much incentive to remedy that.

**Comment by GlenCFL** [comment]
The version of ```@types/atom``` used by atom-typescript is outdated, but I did push new definitions to DefinitelyTyped a few weeks ago with full support for Atom v1.21.

**Comment by DavidLGoldberg** [comment]
Thanks so much for the detailed answers!

I have made all of your changes to package.json, cleared node_modules, `npm install` and tested, and everything is perfect still.  Thanks!  Everything has been converted from coffee -> ES6.  And all but one file needs to be converted to TypeScript.

And yes, no I planned to just leave the spec files as coffee for the foreseeable future.

**Comment by lierdakil** [comment]
@GlenCFL, good to know, I'll go update then. I gather that you would be the first person to pester in case I find something amiss?

**Comment by GlenCFL** [comment]
@lierdakil Sure. Just raise an issue somewhere and mention me.

A bit of a warning though: I plan to convert ```@types/atom``` into an actual external module at some point, hopefully this week. The only global once that happens will be the ```atom``` instance of AtomEnvironment. I've been testing it within [atom-typings](https://github.com/GlenCFL/atom-typings/tree/master/types/atom).

**Comment by smhxx** [comment]
Just to chime in on this, the package that you linked to in your original question (smhxx/atom-ts-transpiler) is actually meant to provide exactly what you seem to be looking for — a way to make Atom "understand" TypeScript without having to transpile everything yourself and commit the output. The issue you linked to was actually caused by Atom's fallback TS handling, which as you said, is kinda wonky at best.

If you're interested, I'd recommend checking it out and seeing if it addresses your needs. A couple other packages that have their source code written in TS have already adopted it, and it seems to be working well for them.

**Comment by chharvey** [comment]
@DavidLGoldberg - Could you rename the title of this issue to something more specific, and put a TLDR at the top of your opening comment? That would help a lot of people (like me) when browsing through the issues. (It would also help SEO.) Thank you!
